File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        vector<int> leftprod(n,1), rightprod(n,1);

        for(int i=1; i<n; i++){
            leftprod[i] = nums[i-1] * leftprod[i-1];
            rightprod[n-i-1] = nums[n-i] * rightprod[n-i];
        }

        vector<int> ans(n);
        
        for(int i=0; i<n; i++){
            ans[i] = leftprod[i]*rightprod[i];
        }
        return ans;

    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        size = len(nums)

        ans = [1] * size

        for i in range(1,size):
            #storing prefix prod in ans
            ans[i] = ans[i-1] * nums[i-1]

        # Multiply by suffix products
        suffix = 1

        for i in range(size-1,-1,-1):
            ans[i] *= suffix
            suffix *= nums[i]

        return ans